src/vm/methodtable.cpp

```diff
@@ -2343,6 +2342,387 @@ SystemVClassificationType MethodTable::ReClassifyField(SystemVClassificationType
+        nestingLevel * 5, "", fieldNum, fieldName, fieldOffset, normalizedFieldOffset, fieldSize, helperPtr->currentUniqueOffsetField,
+        GetSystemVClassificationTypeName(fieldClassificationType),
+        GetSystemVClassificationTypeName(helperPtr->fieldClassifications[helperPtr->currentUniqueOffsetField])));
+
+    helperPtr->currentUniqueOffsetField++;
+    _ASSERTE(helperPtr->currentUniqueOffsetField <= SYSTEMV_MAX_NUM_FIELDS_IN_REGISTER_PASSED_STRUCT);
+
+    return true;
+}
+
+// Assigns the classification types to the array with eightbyte types.
+static void AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHelperPtr helperPtr, unsigned int nestingLevel)
+{
+    // Final classification only happens on the top-level struct.
+    if (helperPtr->inEmbeddedStruct)
+    {
+        return;
+    }
+
+    _ASSERTE(nestingLevel == 0);
+
+    // We're at the top level of the recursion, and we're done looking at the fields.
+    // Now sort the fields by offset and set the output data.
+
+    int sortedFieldOrder[SYSTEMV_MAX_NUM_FIELDS_IN_REGISTER_PASSED_STRUCT];
+    for (unsigned i = 0; i < SYSTEMV_MAX_NUM_FIELDS_IN_REGISTER_PASSED_STRUCT; i++)
+    {
+        sortedFieldOrder[i] = -1;
+    }
+
+    for (unsigned i = 0; i < helperPtr->currentUniqueOffsetField; i++)
+    {
+        _ASSERTE(helperPtr->fieldOffsets[i] < SYSTEMV_MAX_NUM_FIELDS_IN_REGISTER_PASSED_STRUCT);
+        _ASSERTE(sortedFieldOrder[helperPtr->fieldOffsets[i]] == -1); // we haven't seen this field offset yet.
+        sortedFieldOrder[helperPtr->fieldOffsets[i]] = i;
+    }
+
+    // Set the layoutSizes (includes holes from alignment of the fields.)
+    int lastField = -1;
+    for (unsigned i = 0; i < SYSTEMV_MAX_NUM_FIELDS_IN_REGISTER_PASSED_STRUCT; i++)
+    {
+        int ordinal = sortedFieldOrder[i];
+        if (ordinal == -1)
+        {
+            continue;
+        }
+
+        if (lastField == -1)
+        {
+            lastField = ordinal;
+            continue;
+        }
+
+        helperPtr->fieldLayoutSizes[lastField] = helperPtr->fieldOffsets[ordinal] - helperPtr->fieldOffsets[lastField];
+
+        lastField = ordinal;
+    }
+    // Now the last field
+    _ASSERTE(lastField != -1); // if lastField==-1, then the struct has no fields!
+
+    // A field size cannot be bigger than the size of an eightbyte; any necessary padding
+    // will be handled later.
+    _ASSERTE(helperPtr->fieldSizes[lastField] <= SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES);
+    helperPtr->fieldLayoutSizes[lastField] = min(helperPtr->structSize - helperPtr->fieldOffsets[lastField], SYSTEMV_EIGHT_BYTE_SIZE_IN_BYTES);
+
+    // Add logical fields for any padding bytes at the end of the struct.
+    unsigned int paddingOffset = helperPtr->fieldOffsets[lastField] + helperPtr->fieldLayoutSizes[lastField];
```

Comment:
What happens in the following case:
A struct of size 16 bytes with 2 bool fields.
struct S
{

The last field has an offset of 1. The LayoutSize of the last field is 8 (the line just above).